packages/synthetics-sdk-broken-links/src/broken_links.ts

```diff
@@ -139,32 +174,73 @@ export async function runBrokenLinks(
  */
 async function checkOriginLink(
   originPage: Page,
-  options: BrokenLinksResultV1_BrokenLinkCheckerOptions
+  options: BrokenLinksResultV1_BrokenLinkCheckerOptions,
+  startTime: string,
+  total_timeout_millis: number
 ): Promise<BrokenLinksResultV1_SyntheticLinkResult> {
-  // check origin_link
-  const originLinkResult = await checkLink(
-    originPage,
-    { target_uri: options.origin_uri, anchor_text: '', html_element: '' },
-    options,
-    true
-  );
+  let originLinkResult: BrokenLinksResultV1_SyntheticLinkResult;
 
-  try {
-    if (options.wait_for_selector) {
-      await originPage.waitForSelector(options.wait_for_selector, {
-        timeout: options.link_timeout_millis,
-      });
-    }
-  } catch (err) {
-    originLinkResult.link_passed = false;
-    if (err instanceof Error) {
-      originLinkResult.error_type = err.name;
-      originLinkResult.error_message = err.message;
-      process.stderr.write(err.message);
+  // Create all Promise and variables used to set and resolve the time limit
+  // imposed by `total_timeout_millis`
+  let timeLimitTimeout: NodeJS.Timeout;
+  let timeLimitresolver = () => {};
+  const timeLimitPromise = new Promise<boolean>((resolve) => {
+    timeLimitresolver = () => {
+      resolve(false);
+    };
+    const time_used = Date.now() - new Date(startTime).getTime();
+    timeLimitTimeout = setTimeout(
+      timeLimitresolver,
+      total_timeout_millis - time_used - 500
+    );
+  });
+
+  const originLinkResultPromise = async () => {
+    originLinkResult = await checkLink(
+      originPage,
+      { target_uri: options.origin_uri, anchor_text: '', html_element: '' },
+      options,
+      true
+    );
+
+    try {
+      if (options.wait_for_selector) {
+        await originPage.waitForSelector(options.wait_for_selector, {
+          timeout: options.link_timeout_millis,
+        });
+      }
+    } catch (err) {
+      originLinkResult.link_passed = false;
+      if (err instanceof Error) {
+        originLinkResult.error_type = err.name;
+        originLinkResult.error_message = err.message;
+        process.stderr.write(err.message);
+      }
     }
-  }
+    return true;
+  };
+
+  return Promise.race([timeLimitPromise, originLinkResultPromise()]).then(
```

Comment:
why do you `.then` this instead of awaiting the promise.race?

Reply:
in this scenario I need to know whether the originLink correctly finished executing or not so I amusing .then() rather than await. I've changed the other Promise.race() to use .then() as well for consistency